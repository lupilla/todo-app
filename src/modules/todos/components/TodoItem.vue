<template>
  <div class="view" v-show="!isEditing" data-id-test="view">
    <input
      :id="todo.id"
      data-id-test="completed"
      type="checkbox"
      class="toggle"
      aria-label="Check as completed"
      @click="changeComplete(todo)"
      :checked="isChecked"
    />
    <label :for="todo.id" data-id-test="label"> {{ todo.name }}</label>
    <button
      data-id-test="remove"
      type="button"
      class="destroy"
      @click="remove(todo)"
    ></button>
  </div>

  <input
    v-show="isEditing"
    data-id-test="edit"
    class="edit"
    ref="edit"
    type="text"
    aria-label="Edit a Todo"
    @blur="editing(todo)"
    @keyup.enter="editing(todo)"
    @keydown.esc="stopEditing(todo)"
    v-model="name"
  />
</template>
<script>
import { mapActions } from "vuex";
export default {
  name: "TodoItem",
  data() {
    return {
      name: "",
      oldName: "",
    };
  },
  computed: {
    isChecked() {
      return this.todo.completed;
    },
    isEditing() {
      return this.todo.isEditing;
    },
  },
  emits: ["changeStatus", "edit", "stopEdit"],
  props: {
    todo: Object,
  },
  methods: {
    ...mapActions("todos", ["deleteTodo", "changeCompleteTodo"]),
    async changeComplete(todo) {
      this.$emit("changeStatus", todo);
    },
    async editing(todo) {
      if (this.name) {
        try {
          const _todo = { ...todo, name: this.name };
          this.$emit("edit", _todo);
          this.oldName = this.name;
        } catch (error) {
          console.error(error);
        }
      } else {
        try {
          await this.deleteTodo(todo);
        } catch (error) {
          console.error(error);
        }
      }
    },
    async remove(todo) {
      try {
        await this.deleteTodo(todo);
      } catch (error) {
        console.error(error);
      }
    },
    stopEditing(todo) {
      this.$emit("stopEdit", todo);
      this.name = this.oldName;
    },
  },
  mounted() {
    this.name = this.todo.name;
    this.oldName = this.name;
  },
  updated() {
    if (this.todo.isEditing) {
      this.$refs.edit.focus();
    }
  },
};
</script>
<style scoped lang="scss">
.edit {
  position: relative;
  margin: 0;
  width: 100%;
  font-size: 24px;
  font-family: inherit;
  font-weight: inherit;
  line-height: 1.4em;
  border: 0;
  color: inherit;
  padding: 6px;
  border: 1px solid #999;
  box-shadow: inset 0 -1px 5px 0 rgba(0, 0, 0, 0.2);
  box-sizing: border-box;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
.toggle {
  text-align: center;
  width: 40px;
  /* auto, since non-WebKit browsers doesn't support input styling */
  height: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  margin: auto 0;
  border: none; /* Mobile Safari */
  -webkit-appearance: none;
  appearance: none;
  opacity: 0;
}

.toggle:checked + label {
  background-image: url("data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2240%22%20height%3D%2240%22%20viewBox%3D%22-10%20-18%20100%20135%22%3E%3Ccircle%20cx%3D%2250%22%20cy%3D%2250%22%20r%3D%2250%22%20fill%3D%22none%22%20stroke%3D%22%23bddad5%22%20stroke-width%3D%223%22/%3E%3Cpath%20fill%3D%22%235dc2af%22%20d%3D%22M72%2025L42%2071%2027%2056l-4%204%2020%2020%2034-52z%22/%3E%3C/svg%3E");
}

.toggle + label {
  /*
		Firefox requires `#` to be escaped - https://bugzilla.mozilla.org/show_bug.cgi?id=922433
		IE and Edge requires *everything* to be escaped to render, so we do that instead of just the `#` - https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7157459/
	*/
  background-image: url("data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2240%22%20height%3D%2240%22%20viewBox%3D%22-10%20-18%20100%20135%22%3E%3Ccircle%20cx%3D%2250%22%20cy%3D%2250%22%20r%3D%2250%22%20fill%3D%22none%22%20stroke%3D%22%23ededed%22%20stroke-width%3D%223%22/%3E%3C/svg%3E");
  background-repeat: no-repeat;
  background-position: center left;
}
.completed label {
  color: #d9d9d9;
  text-decoration: line-through;
}
.destroy {
  display: none;
  position: absolute;
  top: 0;
  right: 10px;
  bottom: 0;
  width: 40px;
  height: 40px;
  margin: auto 0;
  font-size: 30px;
  color: #cc9a9a;
  margin-bottom: 11px;
  transition: color 0.2s ease-out;
}

.destroy:hover {
  color: #af5b5e;
}

.destroy:after {
  content: "Ã—";
}

.edit {
  display: none;
}
</style>
