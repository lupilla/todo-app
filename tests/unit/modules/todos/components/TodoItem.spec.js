import { shallowMount } from "@vue/test-utils";
import createVuexStore from "../../../mocks/mock-store.js";
import TodoItem from "@/modules/todos/components/TodoItem.vue";
import {
  TODO,
  TODO_NOT_COMPLETED,
  TODO_EDITABLE,
  EDITED_TODO,
} from "../../../mocks/state.js";
const store = createVuexStore();

describe("TodoItem tests", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(TodoItem, {
      global: {
        plugins: [store],
      },
      props: {
        todo: TODO,
      },
    });
  });
  test("TodoItem Component should match with snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("TodoItem Component with isEditing property to false should show view option with edit layout hidden", () => {
    const view = wrapper.find('[data-id-test="view"]');
    const edit = wrapper.find('[data-id-test="edit"]');
    expect(view.isVisible()).toBe(true);
    expect(edit.isVisible()).toBe(false);
  });

  test("TodoItem Component with completed property to false should show a not checked input", () => {
    wrapper = shallowMount(TodoItem, {
      global: {
        plugins: [store],
      },
      props: {
        todo: TODO_NOT_COMPLETED,
      },
    });
    const inputCkecbox = wrapper.find('[data-id-test="completed"]');
    expect(inputCkecbox.element.checked).toBe(false);
  });

  test("TodoItem Component with completed property truthy should show a checked input", () => {
    const inputCkecbox = wrapper.find('[data-id-test="completed"]');
    expect(inputCkecbox.element.checked).toBe(true);
  });

  test("TodoItem Component with isEditing property truthy should show edit input", () => {
    wrapper = shallowMount(TodoItem, {
      global: {
        plugins: [store],
      },
      props: {
        todo: TODO_EDITABLE,
      },
    });
    const view = wrapper.find('[data-id-test="view"]');
    const edit = wrapper.find('[data-id-test="edit"]');
    expect(view.isVisible()).toBe(false);
    expect(edit.isVisible()).toBe(true);
  });

  test("'changeComplete' method should call 'changeComplete' method", async () => {
    const spyCheckTodo = jest.spyOn(wrapper.vm, "changeComplete");
    const inputCkecbox = wrapper.find('[data-id-test="completed"]');
    await inputCkecbox.trigger("click");
    expect(spyCheckTodo).toHaveBeenCalled();
  });
});

describe("Test edit mode", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(TodoItem, {
      global: {
        plugins: [store],
      },
      props: {
        todo: TODO_EDITABLE,
      },
    });
  });
  test("on blur event should call 'editing' method and change 'name' property", async () => {
    const spyEdit = jest.spyOn(wrapper.vm, "editing");
    expect(wrapper.vm.$data.name).toBe("go to the library");
    expect(wrapper.vm.$data.oldName).toBe("go to the library");
    const input = wrapper.find('[data-id-test="edit"]');
    await input.setValue("hola");
    await input.trigger("blur");
    expect(wrapper.vm.$data.name).toBe("hola");
    expect(spyEdit).toHaveBeenCalled();
  });

  test("on enter keypress event should call 'edited' method and change 'name' property", async () => {
    expect(wrapper.vm.$data.name).toBe("go to the library");
    const input = wrapper.find('[data-id-test="edit"]');
    await input.setValue("hola");
    await input.trigger("keypress", {
      key: "enter",
    });
    expect(wrapper.vm.$data.name).toBe("hola");
  });

  test("on enter keypress with 'escape' key event should call 'stopEditing' method and not change 'name' property", async () => {
    const spyStopEdit = jest.spyOn(wrapper.vm, "stopEditing");
    expect(wrapper.vm.$data.name).toBe("go to the library");
    const input = wrapper.find('[data-id-test="edit"]');
    await input.setValue("hola");
    await input.trigger("keydown.esc");
    expect(wrapper.vm.$data.name).toBe("go to the library");
    expect(spyStopEdit).toHaveBeenCalled();
  });

  test("'changeComplete' method should emit 'changeStatus' custom event", async () => {
    const inputCkecbox = wrapper.find('[data-id-test="completed"]');
    await inputCkecbox.trigger("click");
    expect(wrapper.emitted("changeStatus")).toBeTruthy();
  });

  test("'editing' method should emit 'edit' custom event", async () => {
    await wrapper.setData({ name: "hola mundo" });
    await wrapper.vm.editing(TODO);
    expect(wrapper.emitted("edit")).toBeTruthy();
    expect(wrapper.emitted("edit")[0][0]).toEqual(EDITED_TODO);
  });

  test("'stopEditing' method should emit 'stopEdit' custom event", async () => {
    await wrapper.setData({ name: "hola mundo" });
    wrapper.vm.stopEditing(TODO);
    expect(wrapper.emitted("stopEdit")).toBeTruthy();
    expect(wrapper.emitted("stopEdit")[0][0]).toEqual(TODO);
  });
});
