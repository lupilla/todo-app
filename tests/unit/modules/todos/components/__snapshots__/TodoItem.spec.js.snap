// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem tests TodoItem Component should match with snapshot 1`] = `
<div class="view" data-id-test="view"><input id="-N0yGtCoVxtQyt39fNqF" data-id-test="completed" type="checkbox" class="toggle" aria-label="Check as completed"><label for="-N0yGtCoVxtQyt39fNqF" data-id-test="label">go to the library</label><button data-id-test="remove" type="button" class="destroy"></button></div>
<input data-id-test="edit" class="edit" type="text" aria-label="Edit a Todo" style="display: none;">
`;
